File: 2024상반기/5월 2주차/leetcode-783-Minimum Distance Between BST Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void midtour(TreeNode* node, int& prev, int& answer) {
        if (node == nullptr) return;

        midtour(node->left, prev, answer);

        int diff = node->val - prev;
        if (diff < answer) answer = diff;
        prev = node->val;

        midtour(node->right, prev, answer);
    }
    int minDiffInBST(TreeNode* root) {
        int prev = -100001, answer = 100001;
        midtour(root, prev, answer);
        return answer;
    }
};

File: maureen72/Tree/Leetcode-783-MinimumDistanceBetweenBSTNodes.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def minDiffInBST(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        stack = []
        node = root
        #result = 0

        prev = -sys.maxsize
        result = sys.maxsize

        while stack or node:
            while node:
                stack.append(node)
                node = node.left
            node = stack.pop()

            result = min(result, node.val - prev)
            prev = node.val
            node = node.right
        
        return result